# Data Clensing 

1. Finding Unique Values & Analyze NON-Standard Data

In [ ]:
#Get Count of unique values in each columns 

df.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

# 2. Analyze Duplicates

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8525 entries, 0 to 8524
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8525 non-null   object 
 1   Item_Weight                7062 non-null   float64
 2   Item_Fat_Content           8525 non-null   object 
 3   Item_Visibility            8525 non-null   float64
 4   Item_Type                  8525 non-null   object 
 5   Item_MRP                   8525 non-null   float64
 6   Outlet_Identifier          8525 non-null   object 
 7   Outlet_Establishment_Year  8525 non-null   object 
 8   Outlet_Size                6114 non-null   object 
 9   Outlet_Location_Type       8525 non-null   object 
 10  Outlet_Type                8525 non-null   object 
 11  Item_Outlet_Sales          8525 non-null   float64
dtypes: float64(4), object(8)
memory usage: 799.3+ KB


Here we can see, number of totoal records is 8525.

In [ ]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
8520    False
8521    False
8522    False
8523     True
8524     True
Length: 8525, dtype: bool

In [ ]:
df[df.duplicated()==True]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
8523           DRG01        14.80          Low Fat         0.044878   
8524           FDS36         8.38          Regular         0.046982   

         Item_Type  Item_MRP Outlet_Identifier Outlet_Establishment_Year  \
8523   Soft Drinks    75.467            OUT046                      1997   
8524  Baking Goods   108.157            OUT045                      2002   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
8523       Small               Tier 1  Supermarket Type1            765.670  
8524         NaN               Tier 2  Supermarket Type1            549.285

In [ ]:
df[df['Item_Identifier']=='DRG01']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
544            DRG01         14.8          Low Fat         0.044870   
1218           DRG01         14.8          Low Fat         0.044841   
3005           DRG01          NaN          Low Fat         0.044661   
5974           DRG01          NaN          Low Fat         0.078576   
8522           DRG01         14.8          Low Fat         0.044878   
8523           DRG01         14.8          Low Fat         0.044878   

        Item_Type  Item_MRP Outlet_Identifier Outlet_Establishment_Year  \
544   Soft Drinks    76.467            OUT035                      2004   
1218  Soft Drinks    78.367            OUT013                      1987   
3005  Soft Drinks    74.767            OUT027                      1985   
5974  Soft Drinks    78.467            OUT019                      1985   
8522  Soft Drinks    75.467            OUT046                      1997   
8523  Soft Drinks    75.467            OUT046                      1997   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
544        Small               Tier 2  Supermarket Type1           1531.340  
1218        High               Tier 3  Supermarket Type1            918.804  
3005      Medium               Tier 3  Supermarket Type3           2679.845  
5974       Small               Tier 1      Grocery Store            229.701  
8522       Small               Tier 1  Supermarket Type1            765.670  
8523       Small               Tier 1  Supermarket Type1            765.670

In [ ]:
df[df['Item_Identifier']=='FDS36']

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
4986           FDS36         8.38          Regular         0.046878   
5570           FDS36         8.38          Regular         0.046848   
5666           FDS36         8.38          Regular         0.046887   
6299           FDS36         8.38          Regular         0.046960   
8279           FDS36         8.38          Regular         0.047153   
8519           FDS36         8.38          Regular         0.046982   
8524           FDS36         8.38          Regular         0.046982   

         Item_Type  Item_MRP Outlet_Identifier Outlet_Establishment_Year  \
4986  Baking Goods   109.857            OUT035                      2004   
5570  Baking Goods   110.557            OUT013                      1987   
5666  Baking Goods   107.957            OUT046                      1997   
6299  Baking Goods   111.857            OUT049                      1999   
8279  Baking Goods   110.157            OUT017                      2007   
8519  Baking Goods   108.157            OUT045                      2002   
8524  Baking Goods   108.157            OUT045                      2002   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
4986       Small               Tier 2  Supermarket Type1           2306.997  
5570        High               Tier 3  Supermarket Type1           3844.995  
5666       Small               Tier 1  Supermarket Type1           2966.139  
6299      Medium               Tier 1  Supermarket Type1           2966.139  
8279         NaN               Tier 2  Supermarket Type1           3185.853  
8519         NaN               Tier 2  Supermarket Type1            549.285  
8524         NaN               Tier 2  Supermarket Type1            549.285

In [ ]:
df.drop_duplicates(inplace=True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   object 
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), object(8)
memory usage: 865.6+ KB


As we can see count of total number of rows post removing duplicates is 8523 as compared to original row count number which is 8525. 

This means we have are successfull in removing 2 duplicate rows, present in our data.

In [ ]:
df.Item_Fat_Content.unique()

array(['Low Fat', 'Regular', 'low fat', 'LF', 'reg'], dtype=object)

Here we can see for 

* "Low Fat" we have three values, two in different case (Low Fat, low fat) & one with just Initials (LF)  but essential they are same. 

* Similarly we can see for 'Regular' 'reg' which essentially are same.

Lets corret them.. since if we don't our  model/program will consider them as differnt Fat Type and will impact our model performance. 

## Corret Non-Standard Data

In [ ]:
df['Item_Fat_Content']=df['Item_Fat_Content'].replace({'LF':'Low Fat','reg':'Regular','low fat':'Low Fat'})

In [ ]:
df['Item_Fat_Content'].nunique()

2